src/storage/engine/btree_engine.rs

```diff
@@ -0,0 +1,404 @@
+        Ok(self.seek_and_valid(range, false))
+    }
+
+    fn seek_for_prev(&mut self, key: &Key) -> EngineResult<bool> {
+        let range = (Unbounded, Included(key.clone()));
+        Ok(self.seek_and_valid(range, true))
+    }
+
+    fn seek_to_first(&mut self) -> bool {
+        let range = (self.bounds.0.clone(), self.bounds.1.clone());
+        self.seek_and_valid(range, false)
+    }
+
+    fn seek_to_last(&mut self) -> bool {
+        let range = (self.bounds.0.clone(), self.bounds.1.clone());
+        self.seek_and_valid(range, true)
+    }
+
+    #[inline]
+    fn valid(&self) -> bool {
+        debug!("Iterator.valid(): {:?}", self.valid);
+        self.valid
+    }
+
+    fn key(&self) -> &[u8] {
+        assert!(self.valid());
+        debug!("Iterator.key() -> {:?}", unsafe {
+            String::from_utf8_unchecked(self.cur_key.clone().unwrap().to_raw().unwrap())
+        });
+        self.cur_key.as_ref().unwrap().as_encoded()
+    }
+
+    fn value(&self) -> &[u8] {
+        assert!(self.valid());
+        self.cur_value.as_ref().unwrap().as_slice()
+    }
+}
+
+impl Snapshot for BTreeEngineSnapshot {
+    type Iter = BTreeEngineIterator;
+
+    fn get(&self, key: &Key) -> EngineResult<Option<Value>> {
+        self.get_cf(CF_DEFAULT, key)
+    }
+    fn get_cf(&self, cf: CfName, key: &Key) -> EngineResult<Option<Value>> {
+        let tree_cf = self.inner_engine.get_cf(cf);
+        let tree = tree_cf.read().unwrap();
+        let v = tree.get(key);
+        match v {
+            None => Ok(None),
+            Some(v) => Ok(Some(v.clone())),
+        }
+    }
+    fn iter(&self, iter_opt: IterOption, mode: ScanMode) -> EngineResult<Cursor<Self::Iter>> {
+        self.iter_cf(CF_DEFAULT, iter_opt, mode)
+    }
+    #[inline]
+    fn iter_cf(
+        &self,
+        cf: CfName,
+        iter_opt: IterOption,
+        mode: ScanMode,
+    ) -> EngineResult<Cursor<Self::Iter>> {
+        let tree = self.inner_engine.get_cf(cf);
+
+        Ok(Cursor::new(
+            BTreeEngineIterator::new(tree.clone(), iter_opt),
+            mode,
+        ))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct BTreeEngineSnapshot {
+    inner_engine: Arc<BTreeEngine>,
+}
+
+impl BTreeEngineSnapshot {
+    pub fn new(engine: &BTreeEngine) -> Self {
+        Self {
+            inner_engine: Arc::new(engine.clone()),
+        }
+    }
+}
+
+fn write_modifies(engine: &BTreeEngine, modifies: Vec<Modify>) -> EngineResult<()> {
+    for rev in modifies {
+        match rev {
+            Modify::Delete(cf, k) => {
+                let cf_tree = engine.get_cf(cf);
+                cf_tree.write().unwrap().remove(&k);
+            }
+            Modify::Put(cf, k, v) => {
+                let cf_tree = engine.get_cf(cf);
+                cf_tree.write().unwrap().insert(k, v);
+            }
+
+            Modify::DeleteRange(_cf, _start_key, _end_key) => unimplemented!(),
+        };
+    }
+    Ok(())
```

Comment:
Seems this function can still be optimized. Now it iterates the Vec of btrees over and over.